curandommatrices/src/circular.jl

```diff
@@ -0,0 +1,61 @@
+# FIXME: this can be accelerated
+function cplx_phase!(a)
+    i = (blockIdx().x-1) * blockDim().x + threadIdx().x
+    if i <= length(a)
+        @inbounds a[i] = a[i] / sqrt(real(a[i])^2 + imag(a[i])^2)
+    end
+    return nothing
+end
+
+# FIXME: use blocks to support larger than 1024
+
+function _qr_fix!(z::CuMatrix)
+    q, r = CuArrays.qr!(z)
+    ph = diag(r)
+    len = min(length(ph), 1024)
+    @cuda threads=len blocks=16 cplx_phase!(ph)
+    q = CuMatrix(q)
+    idim = size(r, 1)
+    for i=1:idim
+        q[:, i] .*= ph[i]
+    end
+    q[:, 1:idim]
+end
+
+function _qr_fix(z::CuMatrix)
+    a = copy(z)
+    _qr_fix!(a)
+end
+
+function curand(c::COE)
+    z = curand(c.g)
+    u = _qr_fix!(z)
+    transpose(u)*u
+end
+
+function curand(c::CUE)
+    z = curand(c.g)
+    _qr_fix!(z)
+end
+
+function curand(c::CSE)
+    z = curand(c.g)
+    u = _qr_fix!(z)
+    ur = cat([CuMatrix{Float32}([0 -1; 1 0]) for _=1:c.d÷2]..., dims=[1,2])
+    ur*u*ur'*transpose(u)
+end
+
+function curand(c::CircularRealEnsemble)
```

Comment:
Why methods `curand(c::CircularRealEnsemble)`, `curand(c::CircularQuaternionEnsemble)`, `curand(c::HaarIsometry)` have the same implementation? If they are stubs emit a warning, or error. If it is intended comment it or use a `Union` type.

Reply:
I don't feel like they belong in a `Union{...}`, they have little in common actually. Only the sampling implementation looks the same. Looping over types.